### DISCUSSION TITLE
How to run mp_embed_exec_str in nonblocking state.

### DISCUSSION THREAD

**Discussion by nikhil-robinson** [micropython/micropython#13008] [Core Development]
Hello there i was testing the embedded example and the mp_embed_exec_str is blocking if i add an infinite while loop. how to exit out without destining the mpy. i am using esp32 port.

**Comment by jimmo** [comment]
@nikhil-robinson I'm not quite sure what you're asking for -- what do you mean by "how to exit out"?

Also why are you using the embedded example in conjunction with the esp32 port?

**Comment by nikhil-robinson** [comment]
Hi @jimmo  lets say i wrote a program for printing hi  in a while loop 


```
while True:
    print("hI)


```


 if i run this program using the mp_embed_exec_str function  there is no way to stop the program  and the mp_embed_exec_str function blocks the entire code flow  . i want  to exit out of the while loop after some time period , 


i am not using the entire example i am just using the mp_embed_exec_str function from the embedded example . i wanted run python code inside of my application without repl so i took the function mp_embed_exec_str from the embed example and port it to esp32 everything works but if there is a while loop i am unable to exit out of the execution

**Reply by jimmo** [reply]
What do you want to cause the loop to exit?

**Reply by nikhil-robinson** [reply]
lets say i have multiple string of program .

program 1
```python
while True:
    print("hI")
```

program 2
```python
while True:
    print("test1")
```

program3
```python
while True:
    print("hello")
```

i want to execute these programs in sequence with a delay of 2 sec in between and my delay is controlled by an external source in my esp-idf project.

**Reply by jimmo** [reply]
You should make the programs responsible for terminating themselves -- either by providing a method they can call to know when to stop, or they can use `time.ticks_ms` / `time.ticks_diff`.

**Reply by nikhil-robinson** [reply]
for my application this does not apply, the end user will only write the python code and need to update it when ever they want .   My question is how is REPL doing it even if we have infinite while loop sending ctrl+c will stop the program right is it possible to implement the same on my case like an external event will send ctrl+c to the running program ?

**Comment by karfas** [comment]
@nikhil-robinson: If you allow to execute arbitrary python code written by the user, you could
- document that user-supplied code **must** complete within X milliseconds.
- use a watchdog or one of the hardware timers to reboot the device (maybe remember *this* function as faulty and don't execute it again).

However, arbitrary code gives the user a gazillion other ways (besides long-running loops) to brick the device you are building.

**Reply by nikhil-robinson** [reply]
is ther any way to send ctrl +c to the program the same way the repl doees ?

**Reply by karfas** [reply]
Use the source, Luke.

See https://github.com/micropython/micropython/blob/master/py/scheduler.c#L50

I assume you could create a simple user extension calling `mp_sched_exception()` similar to the keyboard interrupt.
